File: benchmark/python/calibrate.py
#!/usr/bin/env python3
"""Edge calibration, Python side -- config P.

The counterpart of rust/src/bin/calibrate.rs. Same fixture, same operations, same
op labels, same JSON shape, so `compare_edges.py` can line the three configs up:

    A  Rust daemon -> BenchHal      (Rust-native plant, no Python in the process)
    B  Rust daemon -> BridgeHal     -> pymocks/sonic_platform   (the deployed path)
    P  Python daemon                -> pymocks/sonic_platform   (this file)

B and P call literally the same mock, so any B-vs-P difference on the platform edge
is the PyO3 crossing and nothing else. A is the no-Python floor.

The DB edge is NOT shared: `XcvrTableHelper::build` is private so Rust is forced onto
its own MockDbTable, and the real swsscommon Python bindings are absent here. Those
rows are measured so the asymmetry is a known quantity rather than a hidden one.

Run via tools/run_calibrate_py.sh so this executes in the same container (and thus
the same CPython) that config B embeds -- calibrating against the host interpreter
would compare two different Pythons.
"""

import argparse
import json
import os
import sys
import time


def bench(label, fn, out):
    """Auto-scaling timer, mirroring the Rust harness: warm up, then grow the
    iteration count until the sample spans >=200ms so short ops still read stably."""
    for _ in range(2000):
        fn()
    n = 2000
    while True:
        t0 = time.perf_counter_ns()
        for _ in range(n):
            fn()
        el = time.perf_counter_ns() - t0
        if el >= 200_000_000 or n >= 2_000_000:
            ns = el / n
            print(f"  {label:<44} {ns:10.1f} ns/call")
            out[label] = ns
            return
        n *= 4


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--fixture", required=True)
    ap.add_argument("--pymocks", required=True)
    ap.add_argument("--num-sfps", type=int, default=32)
    ap.add_argument("--out", default="")
    args = ap.parse_args()

    fixture = os.path.abspath(args.fixture)
    pymocks = os.path.abspath(args.pymocks)

    os.environ["XCVRD_BENCH_FIXTURE"] = fixture
    os.environ["XCVRD_BENCH_NUM_SFPS"] = str(args.num_sfps)
    # Recording costs more than the calls being measured; must stay off here.
    os.environ["XCVRD_BENCH_TRACE"] = "0"
    sys.path.insert(0, pymocks)

    from sonic_platform.platform import Platform
    from mock_swsscommon import Table

    print(f"fixture : {fixture}")
    print(f"pymocks : {pymocks}")
    print(f"num_sfps: {args.num_sfps}")
    print(f"python  : {sys.version.split()[0]}\n")

    chassis = Platform().get_chassis()
    sfp = chassis.get_sfp(0)
    out = {}

    print("config P  (Python daemon -> pymocks, no bridge):")
    bench("num_sfps", chassis.get_num_sfps, out)
    bench("sfp(0)  [handle construction]", lambda: chassis.get_sfp(0), out)
    bench("get_presence  [scalar]", sfp.get_presence, out)
    # Attribute read, not a call: the bridge reads sfp_type via getattr
    # (platform-bridge/src/lib.rs:204), so this measures the same thing B does.
    bench("sfp_type  [string]", lambda: sfp.sfp_type, out)
    bench("get_transceiver_dom_real_value  [27 fields]",
          sfp.get_transceiver_dom_real_value, out)
    bench("get_transceiver_info  [33 fields]", sfp.get_transceiver_info, out)
    bench("get_transceiver_status  [7 fields]", sfp.get_transceiver_status, out)
    bench("call_json(get_transceiver_status_flags)",
          sfp.get_transceiver_status_flags, out)
    bench("read_eeprom(9, 1)", lambda: sfp.read_eeprom(9, 1), out)
    bench("get_change_event(0)", lambda: chassis.get_change_event(0), out)

    print("\nDB edge (mock_swsscommon.Table -- NOT shared with Rust):")
    tbl = Table(None, "TRANSCEIVER_DOM_SENSOR")
    dom = sfp.get_transceiver_dom_real_value()
    fvs = [(k, str(v)) for k, v in dom.items()]
    tbl.set("Ethernet0", fvs)
    bench("db.set  [27 fields]", lambda: tbl.set("Ethernet0", fvs), out)
    bench("db.get", lambda: tbl.get("Ethernet0"), out)
    bench("db.hget", lambda: tbl.hget("Ethernet0", "temperature"), out)
    bench("db.hset", lambda: tbl.hset("Ethernet0", "temperature", "45.0"), out)
    bench("db.get_size_for_key", lambda: tbl.get_size_for_key("Ethernet0"), out)
    bench("db.getKeys", tbl.getKeys, out)

    if args.out:
        with open(args.out, "w") as fh:
            json.dump({"p": out}, fh, indent=2)
        print(f"\nwrote {args.out}")


if __name__ == "__main__":
    main()

File: benchmark/rust/src/bin/calibrate.rs
//! Edge calibration -- measures what the *instrument* costs, per config.
//!
//! Timing the daemon tells you `T_measured = T_orchestration + k * C_edge`. The
//! equivalence gate already yields `k` exactly (it counts every call), so this
//! binary supplies `C_edge` and lets `analyze.py` report a corrected figure plus an
//! explicit error bar. If the correction turns out to be a large fraction of the
//! A/B delta, the result is not publishable -- same halt rule as read amplification.
//!
//!   config A: BenchHal        -- Rust-native plant, no Python in the process
//!   config B: BridgeHal       -- the real PyO3 bridge onto benchmark/pymocks
//!
//! Run:
//!   calibrate --fixture ../fixtures/cmis_40g_lr4.json --pymocks ../pymocks [--config a|b|both]

use std::time::Instant;

use xcvrd_bench::edges::{BenchHal, Fixture};
use xcvrd_rs::hal::{Hal, SfpHandle};

fn arg(name: &str, default: &str) -> String {
    let args: Vec<String> = std::env::args().collect();
    args.iter()
        .position(|a| a == name)
        .and_then(|i| args.get(i + 1))
        .cloned()
        .unwrap_or_else(|| default.to_string())
}

/// Time `f`, auto-scaling the iteration count so short ops still get a stable
/// reading and slow ones (config B is ~17x slower) do not stall the run.
fn bench<F: FnMut()>(label: &str, out: &mut Vec<(String, f64)>, mut f: F) {
    for _ in 0..2_000 {
        f();
    }
    let mut n = 2_000u32;
    loop {
        let t = Instant::now();
        for _ in 0..n {
            f();
        }
        let el = t.elapsed();
        if el.as_millis() >= 200 || n >= 2_000_000 {
            let ns = el.as_nanos() as f64 / n as f64;
            println!("  {label:<44} {ns:10.1} ns/call");
            out.push((label.to_string(), ns));
            return;
        }
        n *= 4;
    }
}

fn exercise(hal: &dyn Hal, tag: &str) -> Vec<(String, f64)> {
    let mut out = Vec::new();
    println!("config {tag}:");

    let sfp: Box<dyn SfpHandle> = hal.sfp(0).expect("sfp(0)");

    bench("num_sfps", &mut out, || {
        std::hint::black_box(hal.num_sfps().unwrap());
    });
    bench("sfp(0)  [handle construction]", &mut out, || {
        std::hint::black_box(hal.sfp(0).unwrap());
    });
    bench("get_presence  [scalar]", &mut out, || {
        std::hint::black_box(sfp.get_presence().unwrap());
    });
    bench("sfp_type  [string]", &mut out, || {
        std::hint::black_box(sfp.sfp_type().unwrap());
    });
    bench("get_transceiver_dom_real_value  [27 fields]", &mut out, || {
        std::hint::black_box(sfp.get_transceiver_dom_real_value().unwrap());
    });
    bench("get_transceiver_info  [33 fields]", &mut out, || {
        std::hint::black_box(sfp.get_transceiver_info().unwrap());
    });
    bench("get_transceiver_status  [7 fields]", &mut out, || {
        std::hint::black_box(sfp.get_transceiver_status().unwrap());
    });
    bench("call_json(get_transceiver_status_flags)", &mut out, || {
        std::hint::black_box(sfp.call_json("get_transceiver_status_flags").unwrap());
    });
    bench("read_eeprom(9, 1)", &mut out, || {
        std::hint::black_box(sfp.read_eeprom(9, 1).unwrap());
    });
    bench("get_change_event(0)", &mut out, || {
        std::hint::black_box(hal.get_change_event(0).unwrap());
    });
    out
}

fn to_map(rows: &[(String, f64)]) -> serde_json::Value {
    let m: serde_json::Map<String, serde_json::Value> = rows
        .iter()
        .map(|(k, v)| (k.clone(), serde_json::Value::from(*v)))
        .collect();
    serde_json::Value::Object(m)
}

fn main() {
    let fixture_path = arg("--fixture", "../fixtures/cmis_40g_lr4.json");
    let pymocks = arg("--pymocks", "../pymocks");
    let which = arg("--config", "both");
    let num_sfps: usize = arg("--num-sfps", "32").parse().expect("--num-sfps");

    // Set before ANY GIL acquire: pyo3's auto-initialize starts CPython lazily on the
    // first `Python::with_gil`, and sys.path is fixed at interpreter startup, so
    // exporting this later would silently import the real plugin (or fail).
    let abs = std::fs::canonicalize(&pymocks).unwrap_or_else(|e| panic!("--pymocks {pymocks}: {e}"));
    let fx_abs = std::fs::canonicalize(&fixture_path)
        .unwrap_or_else(|e| panic!("--fixture {fixture_path}: {e}"));
    std::env::set_var("PYTHONPATH", &abs);
    std::env::set_var("XCVRD_BENCH_FIXTURE", &fx_abs);
    std::env::set_var("XCVRD_BENCH_NUM_SFPS", num_sfps.to_string());
    // Tracing must stay off here: recording costs more than the calls being measured.
    std::env::set_var("XCVRD_BENCH_TRACE", "0");

    println!("fixture : {}", fx_abs.display());
    println!("pymocks : {}", abs.display());
    println!("num_sfps: {num_sfps}\n");

    let fx = Fixture::from_path(fx_abs.to_str().unwrap()).expect("load fixture");

    let mut results = serde_json::Map::new();

    if which == "a" || which == "both" {
        let hal = BenchHal::new(fx.clone(), num_sfps);
        let a = exercise(&hal, "A  (BenchHal -- Rust-native, no Python)");
        results.insert("a".into(), to_map(&a));
        println!();
    }

    if which == "b" || which == "both" {
        match xcvrd_rs::hal::BridgeHal::new() {
            Ok(hal) => {
                let b = exercise(&hal, "B  (BridgeHal -- real PyO3 onto pymocks)");
                results.insert("b".into(), to_map(&b));
            }
            Err(e) => {
                // Loud, not fatal: config A's numbers are still valid on their own.
                eprintln!("config B unavailable: {e}");
                eprintln!("  (needs libpython + an importable sonic_platform on PYTHONPATH)");
            }
        }
    }

    // Per-op A->B ratio: this IS the PyO3 marshalling tax, per call, by operation.
    if let (Some(a), Some(b)) = (results.get("a"), results.get("b")) {
        println!("\nPyO3 tax (B / A), per operation:");
        let (a, b) = (a.as_object().unwrap(), b.as_object().unwrap());
        for (k, va) in a {
            if let (Some(va), Some(vb)) = (va.as_f64(), b.get(k).and_then(|v| v.as_f64())) {
                println!("  {k:<44} {:8.1}x   (+{:.0} ns)", vb / va, vb - va);
            }
        }
    }

    let out = arg("--out", "");
    if !out.is_empty() {
        std::fs::write(&out, serde_json::to_string_pretty(&results).unwrap())
            .unwrap_or_else(|e| panic!("--out {out}: {e}"));
        println!("\nwrote {out}");
    }
}
